docs: add blog post for js-ipfs 0.44.0 (#414)
ipfs/blog: content/post/094-js-ipfs-0.44.md

---

> date: 2020-05-18
> url: 2020-05-18-js-ipfs-0-44

Reviewer: ```suggestion
date: 2020-05-20
url: 2020-05-20-js-ipfs-0-44
```
Update launch date?

title: js-ipfs 0.44.0 released
header_image: js-ipfs-placeholder.png
author: Alex Potsides
---

# 🔦 Highlights

> Cancellable requests and a new browser datastore

> `js-IPFS@0.44.0` is out of the door with support for cancelling requests and a leaner, meaner datastore for the browser!

Reviewer 2: Could use more rhyming, but this'll do.

Author: Cleaner?

## 🤖 Cancellable requests

> A user should be able to cancel long-lived API operations. For example, if you are fetching the content for a [CID][] from the network, and that CID is not resolveable, you should be able to set a timeout value for the request, after which the content would no longer be sought and control would be returned to your code with an error message describing what happened.

Reviewer 3: ```suggestion
A user should be able to cancel long-lived asynchronous API operations. For example, if you are fetching the content for a [CID][] from the network, and that CID is not resolveable, you should be able to set a timeout value for the request, after which the content would no longer be sought and control would be returned to your code with an error message describing what happened.
```

> A user should be able to cancel long-lived asynchronous API operations. For example, if you are fetching the content for a [CID][] from the network, and that CID is not resolveable, you should be able to set a timeout value for the request, after which the content would no longer be sought and control would be returned to your code with an error message describing what happened.

Reviewer 2: CID needs URL

Author: Done!

This is not as straightforward as it seems in JavaScript because a [Promise](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise) is returned from the API call to resolve a piece of content and the [Promises spec](https://promisesaplus.com/) includes nothing about cancelling a Promise.

> However, the browser [fetch API](https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API) has a notion of an [AbortSignal](https://developer.mozilla.org/en-US/docs/Web/API/AbortSignal) which can be used to abort web requests, they are interacted with via an [AbortController](https://developer.mozilla.org/en-US/docs/Web/API/AbortController):

Reviewer 3: ```suggestion
However, the browser [fetch API](https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API) has a notion of an [AbortSignal](https://developer.mozilla.org/en-US/docs/Web/API/AbortSignal) which can be used to abort web requests. They are interacted with via an [AbortController](https://developer.mozilla.org/en-US/docs/Web/API/AbortController):
```

```javascript
const controller = new AbortController()

setTimeout(() => {
  controller.abort()
}, 1000)

try {
  await result = fetch('http://example.com', {
    signal: controller.signal
  })
} catch (err) {
  console.err(err) // prints: user aborted the request
}
```

We’ve taken this approach and integrated it into every API call, so you can now use AbortControllers to cancel requests! We’ve also used this mechanism to add a `timeout` option to every API call which internally uses an AbortSignal to stop the request after the specified amount of time:

```javascript
const cid = new CID('QmWillNeverResolve')

try {
  await ipfs.get(cid, {
    timeout: 1000 // abort after 1000ms
  })
} catch (err) {
  console.err(err)

> }

Reviewer 2: Can you clarify via comment or otherwise that adding a timeout *does* throw an exception? (like in prev example)

Author: Done!

```

You could also roll this yourself:

```javascript
const cid = new CID('QmWillNeverResolve')
const controller = new AbortController()

setTimeout(() => {
  controller.abort()
}, 1000)

try {
  await ipfs.get(cid, {
    signal: controller.signal
  })
} catch (err) {
  console.err(err)
}
```

The AbortSignal is passed all the way down the call stack to the libp2p and ipld components that underpin IPFS. In this initial release, timeouts are supported by the IPFS layer and in future releases libp2p and ipld will use the AbortSignal to also free up resources and perform any other necessary cleanup that would have been done if the request had completed successfully.

## New browser datastore

> `js-IPFS@0.44.0` brings a new datastore to the browser. In the browser all blocks and other repo data is stored in [IndexedDB](https://developer.mozilla.org/en-US/docs/Web/API/IndexedDB_API) as at the time of writing it is the only way available to persist and query large amounts of user data.

Reviewer 3: ```suggestion
`js-IPFS@0.44.0` brings a new datastore to the browser. In the browser, all blocks and other repo data is currently stored in [IndexedDB](https://developer.mozilla.org/en-US/docs/Web/API/IndexedDB_API), given that it is the only way available to persist and query large amounts of user data at the time of writing.
```

Reviewer 3: I think that my suggestion is more clear

> `js-IPFS@0.44.0` brings a new datastore to the browser. In the browser, all blocks and other repo data is currently stored in [IndexedDB](https://developer.mozilla.org/en-US/docs/Web/API/IndexedDB_API), given that it is the only way available to persist and query large amounts of user data at the time of writing.

Author: Do we not?

> `js-IPFS@0.44.0` brings a new datastore to the browser. In the browser, all blocks and other repo data is currently stored in [IndexedDB](https://developer.mozilla.org/en-US/docs/Web/API/IndexedDB_API), given that it is the only way available to persist and query large amounts of user data at the time of writing.

Reviewer 2: Until we request the permission and call .persist(), etc it can be blown away at any point, even if some other origin hits the overall browser-level quota: ipfs/js-ipfs#2937

> `js-IPFS@0.44.0` brings a new datastore to the browser. In the browser, all blocks and other repo data is currently stored in [IndexedDB](https://developer.mozilla.org/en-US/docs/Web/API/IndexedDB_API), given that it is the only way available to persist and query large amounts of user data at the time of writing.

Author: I think the difference is largely semantic.  Anyway I didn't want to keep repeating the word 'store' in the sentence.

In node and go-ipfs, [LevelDB](https://en.wikipedia.org/wiki/LevelDB) has been the datastore of choice for application data (though blocks have been stored on the filesystem) and go-IPFS is moving to [Badger](https://dgraph.io/blog/post/badger/)).

> Out of this grew [interface-datastore](https://github.com/ipfs/interface-datastore)—a spec for storing key/value pairs implemented by the datastores used by IPFS. In the browser this was [datastore-level](https://github.com/ipfs/js-datastore-level) which was backed by [level-js](https://www.npmjs.com/package/level-js), which was backed by IndexedDB.

Reviewer 2: s/—/ — /

Author: Done!

Phew.

> To remove a few layers here, we’ve created the [datastore-idb](https://github.com/ipfs/js-datastore-idb) which implemenets the interface-datastore spec, backed by IndexedDB without going via level.

Reviewer 2: typo: implements

Author: Done!

What does this mean?

The new datastore is smaller and faster and the upgrade is seamless since it’s still IndexedDB under the hood:

```
datastore-idb up to 44% faster and 69% smaller

batch idb x 5.26 ops/sec ±4.57% (29 runs sampled)
batch level x 2.92 ops/sec ±3.76% (19 runs sampled)
Fastest is batch idb

size 37.32KB to 11.41KB
```

## ✨New features

* Cancellable API calls ([#2993](https://github.com/ipfs/js-ipfs/issues/2993)) ([2b24f59](https://github.com/ipfs/js-ipfs/commit/2b24f590041a0df9da87b75ae2344232fe22fe3a)), closes [#3015](https://github.com/ipfs/js-ipfs/issues/3015)
* `ipfs.block.put` now accepts the `pin` argument in line with go-ipfs [#3015](https://github.com/ipfs/js-ipfs/issues/3015)

## 🦟 Bugs fixed

* Fixes browser script tag example ([#3034](https://github.com/ipfs/js-ipfs/issues/3034)) ([ee8b769](https://github.com/ipfs/js-ipfs/commit/ee8b769b96f7e3c8414bbf85853ab4e21e8fd11c)), closes [#3027](https://github.com/ipfs/js-ipfs/issues/3027)

> * No longer produces browser bundle with all IPLD formats in the browser ([#3025](https://github.com/ipfs/js-ipfs/issues/3025)) ([e6079c1](https://github.com/ipfs/js-ipfs/commit/e6079c17d5656e92dd5191f0581000c6a782c7ed))

Reviewer 2: Does this reduce bundle size? should mention, if so

Author: No, it's just an extra file that wasn't used.

* Remove node globals ([#2932](https://github.com/ipfs/js-ipfs/issues/2932)) ([d0d2f74](https://github.com/ipfs/js-ipfs/commit/d0d2f74cef4e439c6d2baadba1f1f9f52534fcba))
* `typeof` bug when passing timeout to dag.get ([#3035](https://github.com/ipfs/js-ipfs/issues/3035)) ([026a542](https://github.com/ipfs/js-ipfs/commit/026a5423e00992968840c9236afe47bdab9ee834))

> * Removes use of node globals to prepare us for when browser bundlers stop automatically including them [#2932](https://github.com/ipfs/js-ipfs/pull/2932)

Reviewer 4: ```suggestion
* Removes use of node globals and built ins to prepare us for when browser bundlers stop automatically including them [#2932](https://github.com/ipfs/js-ipfs/pull/2932)
```

> * Removes use of node globals and built ins to prepare us for when browser bundlers stop automatically including them [#2932](https://github.com/ipfs/js-ipfs/pull/2932)

Reviewer 2: two entries for node globals removal. maybe link "browser bundlers stop automatically including them" to one of the stories or announcements about it?

Author: Done!
